### ISSUE TITLE
Critical - almost 50% of keywords not being indexed

### ISSUE THREAD

**Issue by rkyoku** [teamtnt/tntsearch#145]
Hi there,

I have a thousand products with title, description, and tags.

Two of them have the "converter" keyword in their title. (EDIT : see below, I found out that there are actually almost 50% of the words that are not being indexed)

However, this keyword never gets indexed. There are other keywords indexed like "convert", "convers", ... but not "converter".

Could you please tell me what I'm doing wrong, and how I could get it indexed when creating the index?

Cheers,

**EDIT:** if it helps, here are the full titles:
"Operating Lease Converter Excel Model"
"Binary Number Converter Download Spreadsheet"

**Comment by ultrono** [comment]
Do the words appear within the doc list table of the sqllite file?

**Comment by rkyoku** [comment]
No they don't appear in the doc list.

I performed this query: `SELECT * FROM wordlist WHERE term like :keyword ORDER BY num_hits DESC`, with keyword being `'conv%'`, to check... But the "converter" keyword is not appearing in the result set.

(Hence, it does not matter much what kind of TNT search I am performing, since the keyword is not indexed :) )

**Comment by rkyoku** [comment]
I just performed a search for all indexed words (`SELECT count(*) FROM wordlist`), and the result is **5346**.

Then I used a tool for counting the unique keywords myself, and I found **10099** words (including stopwords).

Maybe stopwords like "of" "or" etc. are not indexed by TntSearch (I don't know, I'm supposing), but in any case, there aren't 5 thousand unique stopwords in my text (50% of the whole text...)

The amount of indexed products is good, though.

Any help on that one? I cannot put my code live if 50% of the keywords will be missing from the searches :(

**Comment by stokic** [comment]
You actually haven't provided enough info to discover if there is a problem or if you are just misinterpreting the index. For one thing, indexer uses a stemmer so it "stems" the word to it's root (don't want to go into too much detail, you can Google about stemming if interested). 

To quote Wikipedia "A stemmer for English, for example, should identify the string "cats" (and possibly "catlike", "catty" etc.) as based on the root "cat", so it's very possible and frankly desirable that the index has much less stems than you have unique words in your vocabulary.

That being said, what exactly is your database/table structure?
Can you show us the query you are using for indexing?
Can you show your TNT Search settings?
Did you try reindexing?

**Comment by rkyoku** [comment]
The database would be too much to put here (and maybe confidential).

Here is the query I use for building the index (which partly answers your above question):
```
SELECT p.id, p.title,
		pvi1.data AS desc_short, pvi2.data AS desc_long,
		m.pseudo AS author_name,
		GROUP_CONCAT(t.label SEPARATOR " ") AS tags
FROM product AS p
INNER JOIN member AS m 
	ON m.id = p.author_id
INNER JOIN product_version AS pv
	ON pv.id = p.current_version_id
INNER JOIN product_version_info AS pvi1
	ON pvi1.product_version_id = pv.id
	AND pvi1.info = "desc_short"
INNER JOIN product_version_info AS pvi2
	ON pvi2.product_version_id = pv.id
	AND pvi2.info = "desc_long"
LEFT JOIN product_tag AS pt
	ON pt.product_id = p.id
LEFT JOIN tag AS t 
	ON t.id = pt.tag_id
WHERE p.is_active = 1 AND p.is_published = 1 AND p.current_version_id IS NOT NULL
GROUP BY p.id
```

My TNT search settings are not modified, except for:
```
$this->_tnt->fuzziness = true;
$this->_tnt->fuzzy_prefix_length = 1; // Mandatory, otherwise it won't find a typo like "dwonload"
```

I tried reindexing.

I am not acustomed to stemming, but bottom line, IMHO my "converted" word should be searchable.

I'd be glad to do whatever you need to make it work! Maybe there is an option to modify this stemmer so that it is less agressive?

Cheers,

**Comment by stokic** [comment]
What is your fuzzy_distance setting? 

If the stem of the word "converted" is indexed, then with most settings it'll be searchable. Are you sure that the query gets this info ie. does the query result have this description with the word "converted"? Are other words from the same "line" indexed?

**Comment by stokic** [comment]
Btw your examples:

http://tntstudio.hr:8081/e234e

with you settings:

http://tntstudio.hr:8081/89211

results:

http://tntstudio.hr:8081/5c9b2

So everything seems fine (even if you search "converteD", because of stemming and fuzziness it'll be found)...

**Comment by nticaric** [comment]
It's also good to mention that by default TNTSearch is case sensitive, so if you have "Converter" in your db, and search for "converter" you won't get anything :)

Btw. the stemmed version of converter is `convert` so you should find that in your wordlist table.

**Comment by rkyoku** [comment]
Yeah I have `convert` in my list (as stated in my first post) but it is negative for me not to be able to find `converter`.

Is there a way to modify the stemmer behavior?

**Comment by rkyoku** [comment]
> What is your fuzzy_distance setting?
I did not modify it :-/

> by default TNTSearch is case sensitive
I did not see this behavior. Both searches for `Convert` and `convert` gave me 49 results.

> It's also possible to disable the stemmer if you don't need it.
Oooooh! Right, maybe that's indeed what I need!! I'll look into it, thanks :)

**Comment by rkyoku** [comment]
Ok now, when using no stemmer, I get **8072 indexed words** against the **10099 unique words** I found using a tool. It's way better than the **5346 indexed words** I had earlier.

I understand that most people may not need this, but for those who need to further sort the results (again, because of ticket #93 ), I think that the stemmer gets in the way.

For those who need it, I used this code:
```
$oIndex = $this->_tnt->createIndex($sModel);
$oIndex->setStemmer(new \TeamTNT\TNTSearch\Stemmer\NoStemmer());
$oIndex->query($sReq);
//$oIndex->setLanguage('german');
$oIndex->run();
```

Thanks @stokic and @nticaric for pointing this feature to me! (maybe this should be added to the readme.md though)

I'm closing the ticket, thanks again!
